libsolidity/codegen/ir/IRGenerationContext.cpp

```diff
@@ -153,10 +132,48 @@ string IRGenerationContext::trySuccessConditionVariable(Expression const& _expre
 	return "trySuccessCondition_" + to_string(_expression.id());
 }
 
-string IRGenerationContext::internalDispatch(size_t _in, size_t _out)
+string IRGenerationContext::generateInternalDispatchFunction(Arity const& _arity)
+{
+	// UNIMPLEMENTED: Internal library calls via pointers are not implemented yet.
+	// We're not generating code for internal library functions here even though it's possible
+	// to call them via pointers. Right now such calls end up triggering the `default` case in
+	// the switch above.
+	set<FunctionDefinition const*> functions;
+	for (auto const& contract: mostDerivedContract().annotation().linearizedBaseContracts)
+		for (FunctionDefinition const* function: contract->definedFunctions())
+			if (
+				FunctionType const* functionType = TypeProvider::function(*function)->asCallableFunction(false);
+				!function->isConstructor() &&
+				TupleType(functionType->parameterTypes()).sizeOnStack() == _arity.in &&
+				TupleType(functionType->returnParameterTypes()).sizeOnStack() == _arity.out
+			)
+			{
+				functions.insert(function);
+				enqueueFunctionForCodeGeneration(*function);
+			}
+
+	return generateInternalDispatchFunction(_arity, functions);
+}
+
+string IRGenerationContext::generateInternalDispatchFunction(Arity const& _arity, set<FunctionDefinition const*> const& _functions)
 {
-	string funName = "dispatch_internal_in_" + to_string(_in) + "_out_" + to_string(_out);
-	return m_functions.createFunction(funName, [&]() {
+	vector<map<string, string>> cases;
+	for (auto const* function: _functions)
+	{
+		solAssert(function, "");
+		solAssert(getFunctionArity(*function) == _arity, "A single dispatch function can only handle functions of one arity");
+		solAssert(!function->isConstructor(), "");
+		// 0 is reserved for uninitialized function pointers
+		solAssert(function->id() != 0, "Unexpected function ID: 0");
+
+		cases.emplace_back(map<string, string>{
+			{"funID", to_string(function->id())},
+			{"name", buildFunctionName(*function)}
+		});
+	}
```

Comment:
See https://github.com/ethereum/solidity/pull/8797#discussion_r425235038

Reply:
If you call `internalDispatch` twice, it should not do any work on the second call.

Reply:
OK. That's a good point. I'll move it in then. But there will still be a move, just with `_functions` or we risk having dangling references.

Reply:
Why do you think there could be dangling references? The lambda function is either discarded or called directly.

Reply:
The lambda would iterate over `_functions` which is passed as a `const` reference from the outside. If `generateInternalDispatchFunction()` gets called by its other overload, the reference points at its local variable called `functions`. That variable no longer exists when the lambda eventually gets called.

In #8797 this changes and the the reference will be pointing at `IRGenerationContext::m_internalDispatchMap` instead which we know will not go out of scope but it's just a happy coincidence - we don't do anything specific to ensure that.

Reply:
This part is now in #8951.
- The loop is now back inside the lambda.
- As for moving `_functions` into the lambda, let's continue the discussion in https://github.com/ethereum/solidity/pull/8951#discussion_r425352387